src/test/run-pass/ctfe/const-fn-type-name.rs

```diff
@@ -0,0 +1,37 @@
+// run-pass
+
+#![feature(core_intrinsics)]
+#![feature(const_fn)]
+#![allow(dead_code)]
+
+const fn type_name_wrapper<T>(_: &T) -> &'static str {
+    unsafe { core::intrinsics::type_name::<T>() }
```

Comment:
Why does this even compile, without the intrinsic being marked as `const`?

Reply:
I'm not sure what you mean? With `type_name` *not* added to `emulate_intrinsic()`, you get I believe that message about "only a curated set of intrinsics are allowed in const fn" or something like that.

Once added, as this PR does, it works like all the other ones do.

Reply:
I believe @RalfJung's question is why it didn't need to be added to https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/qualify_consts.rs#L465-L488 which just boils down to you can't do

```rust
const fn foo<T>() -> &'static &'static str {
    unsafe { &core::intrinsics::type_name::<T>() }
}
```

while

```rust
const fn foo<T>() -> &'stati usize {
    unsafe { &core::intrinsics::size_of::<T>() }
}
```

is allowed.

I think we can just scrap that list and just forbid or allow all intrinsics, because you won't be able call intrinsics directly anyway (except for transmute, which has https://github.com/rust-lang/rust/blob/da9ebc828c982d2ed49396886da85011e1b0a6c0/src/librustc_mir/transform/qualify_consts.rs#L1211)

Reply:
`emulate_intrinsic` is the dynamic part, implementing the functionality.

But shouldn't there also somewhere be a static part that prevents non-emulated intrinsics from getting called in a `const`?

Reply:
> I think we can just scrap that list and just forbid or allow all intrinsics, because you won't be able call intrinsics directly anyway (except for transmute

Uh... that would need commitment from `@rust-lang/libs` to never reexport any more intrinsics directly. `transmute` used to not be the only one, `copy_nonoverlapping` and `copy` also used to be directly accessible. And then there are all the SIMD intrinsics, I am not sure if these all have wrappers or get reexported (Cc @gnzlbg)

Also not sure why your test functions return `&intrinsic()` instead of just `intrinsic`? They still need to be a `const fn`? What does this have to do with promotion?

Ideally we'd declare them as `const fn` and maybe add `#[rustc_promotable]` instead of that list -- i.e., the signature in the `extern` block could determine such things.

Reply:
> Uh... that would need commitment from @rust-lang/libs to never reexport any more intrinsics directly. 

we can just forbid them ;) There is nothing exposed that users can access from stable.

> They still need to be a const fn? What does this have to do with promotion?

The big list just checks for promotability, which happens to be a subset of const evaluability.

> Ideally we'd declare them as const fn and maybe add #[rustc_promotable] instead of that list -- i.e., the signature in the extern block could determine such things.

I have tried once to enable `const` on extern blocks's `fn` signatures, it wasn't fun, but certainly doable.

Reply:
> But shouldn't there also somewhere be a static part that prevents non-emulated intrinsics from getting called in a const?

There should, and I believed it to be that list, and it once even was that list. Looks like a refactoring gone wrong.

Reply:
> we can just forbid them ;) There is nothing exposed that users can access from stable.

I don't know how you can be so sure of this. `core::intrinsics::copy` is available on stable even though it was never meant to.

> The big list just checks for promotability, which happens to be a subset of const evaluability.

Uh why do we promote all these things, I thought we'd not promote new things any more? And where is the check for const-callability?

> There should, and I believed it to be that list, and it once even was that list. Looks like a refactoring gone wrong.

Ouch, so we don't check for const-callability any more? That's weird though, intrinsics are not `const fn` so they must be explicitly allowed somewhere?

Reply:
> Ouch, so we don't check for const-callability any more? That's weird though, intrinsics are not `const fn` so they must be explicitly allowed somewhere?

Whether or not you can call an intrinsic within a `const fn` is, as far as I can tell, determined [entirely here.](https://github.com/rust-lang/rust/blob/7096ff0ce16b0544b717986ec335798b3151dd8e/src/librustc_mir/const_eval.rs#L370)

Which is to say, the presence of a match arm for a given intrinsic within `emulate_intrinsic()` is all that matters, which is why the changes in my PR make my test file compile successfully and work properly.

Reply:
This problem is independent of this PR, we should not further dwell on it. I have written a short explanation of the problem to @vertexclique and they'll have a look.

Reply:
> This problem is independent of this PR,

Is there an issue tracking this ? I expect that calling a non-`const fn` from within a `const fn` always results in a compilation error, and most of the intrinsics are not `const fn`. 

This is a different issue from `miri` (the tool) being able to run Rust code that calls these intrinsics at run-time or not. 

The only thing I can imagine is going on here is that `unsafe` inside `const fn` does not only allow calling `unsafe const fn`s, but also non-`const fn` from within `const fn`.

Reply:
> Is there an issue tracking this ? I expect that calling a non-`const fn` from within a `const fn` always results in a compilation error, and most of the intrinsics are not `const fn`.

You never call the _real_ intrinsics from `const fn`s though. You call the `emulate_intrinsics()` version, if it exists. If it doesn't exist, you can't call that intrinsic at all in a `const fn`.

Reply:
What do you mean by "the `emulate_intrinsics` version"? Your example is calling the `core::intrinsics::type_name` intrinsics, which has type [`pub unsafe extern "rust-intrinsic" fn type_name<T: ?Sized>() -> &'static str`](https://github.com/rust-lang/rust/blob/master/src/libcore/intrinsics.rs#L686). Therefore, independently of whether miri can evaluate it or not, calling it within a `const fn` should always fail to type check.

This PR does not change the type of that intrinsic to `pub unsafe extern "rust-intrinsic" const fn type_name<T: ?Sized>() -> &'static str`, so either `unsafe {}` allows calling non-`const fn`s during constant evaluation (which needs a tracking issue), or there is here a different feature at play that "magically" allows changing the type of functions  during constant evaluation, e.g., from `fn` to `const fn` in this case, which would need a tracking issue as well, and then tracking issues for the functions for which this happens.

Reply:
> Your example is calling the `core::intrinsics::type_name` intrinsics

Which, like all intrinsics, is not a real function and is constructed directly by the compiler in either `librustc_codegen_llvm` or `librustc_mir` depending on whether or not the "call" to it is done from inside a `const fn`.

Again, in that case, I'm _not_ calling it, because I can't, because there is not _currently_ a match arm for `type_name` in `emulate_intrinsic()`.

It may indeed make more _sense_ if the publicly visible "function signatures" for intrinsics expressed that they were all "const", but in a practical/technical sense it doesn't actually matter because it's impossible for them to be anything _other_ than "const", being that they have their implementations explicitly constructed in all cases.

Reply:
> Which, like all intrinsics, is not a real function and is constructed directly by the compiler 

How intrinsics are implemented in the language today is an implementation detail that shouldn't leak to users. From the user's POV, an intrinsic is just a normal Rust function that they can call, and if a non-`const fn` can be called from a `const fn`, then that's IMO a soundness bug.

> but in a practical/technical sense it doesn't actually matter because it's impossible for them to be anything other than "const", 

Ehm.. we have thousands of stable intrinsics available in Rust (and ca. 5000 available in nightly Rust), many of which have side-effects and cannot, therefore, be `const fn`.

Reply:
> You never call the "real" librustc_codegen_llvm intrinsics from const fns though. You call the emulate_intrinsics() version from librustc_mir, if it exists. If it doesn't exist, you can't call that intrinsic at all in a const fn in an actually-const context.

The problem is that you get the error too late: you only get an error from a missing implementation on `emulate_intrinsics` when actually running the code, but you should get an error when type-checking. This makes a difference e.g. in `const fn` (that might not get used in a crate but only exported), or in generic consts (once they exist).

I will open an issue, because that problem is not changed by this PR.

Reply:
Reported as https://github.com/rust-lang/rust/issues/61495.